### PR TITLE
Implement tracking of referrals

### PR DESCRIPTION
**Pull request by lukebarnard1**
This also modifies (or fixes) auto-joining. What remains is to join rooms if the team token is specified in the query parameters (so the token does not need GETing from RTS).

### CODE REVIEW

File: src/components/structures/MatrixChat.js [hunk 1]
```diff
@@ -1061,6 +1061,7 @@ module.exports = React.createClass({
                     defaultIsUrl={this.getDefaultIsUrl()}
                     brand={this.props.config.brand}
                     teamServerConfig={this.props.config.teamServerConfig}
+                    queryParams={this.props.startingFragmentQueryParams}
```
Comment (dbkr) [inline src/components/structures/MatrixChat.js]: If we're just extracting one thing from this, I'd prefer to extract it in MatrixChat and pass the actual thing in like we do for email. That or we should just pass the query params in and have Registration extract everything, but ideally query params should be handled by the containing web page, not the components, so failing that I'd rather keep them confined to MatrixChat.

File: src/components/structures/login/Registration.js [hunk 2]
```diff
@@ -209,24 +210,43 @@ module.exports = React.createClass({
...
+            ) {
+                // Track referral, get team_token in order to retrieve team config
+                self._rtsClient.trackReferral(
+                    self.props.queryParams.referrer,
+                    response.user_id,
+                    self.state.formVals.email
+                ).then((args) => {
+                    const teamToken = args[1].team_token;
+                    // Store for use /w welcome pages
+                    window.localStorage.setItem('mx_team_token', teamToken);
```
Comment (lukebarnard1) [inline src/components/structures/login/Registration.js]: I hadn't considered what happens when `mx_team_token` gets removed from local storage. Riot would not consider them part of the team at that point. We might need another endpoint on the RTS to be hit on login to re-verify team membership and send the `team_token` again.

File: src/RtsClient.js [hunk 3]
```diff
@@ -12,4 +12,21 @@ export default class RtsClient {
...
         });
     }
+
+    trackReferral(referrer, user_id, user_email) {
+        return request({
+            url: this._url + '/register',
+            json: true,
+            body: {referrer, user_id, user_email},
+            method: 'POST',
+        });
```
Comment (lukebarnard1) [inline src/RtsClient.js]: The team server looks each email up with the identity server to check that it's a match to the given `user_id`. Any `user_id`/`user_email` pair can only be "referred" once. The email has to be a team email so unless someone manages to genuinely sign people up from their uni, the referral won't count.

File: src/components/structures/MatrixChat.js (hunk 1 above)
Comment (lukebarnard1) [inline src/components/structures/MatrixChat.js]: Works for me. `Registration` will need the `team_token` later, but that could just be another prop

File: src/RtsClient.js (hunk 3 above)
Comment (dbkr) [inline src/RtsClient.js]: OK, that should be fine then.

File: src/components/structures/login/Registration.js (hunk 2 above)
Comment (dbkr) [inline src/components/structures/login/Registration.js]: Had a chat about what we could do here & concluded that yep, we just need to check with the team server on login to see what team a newly logged in user is in, if any.

**Comment by dbkr** [comment]
OK, yeah, otherwise lgtm

**Pull request merged**